addons/mail/static/src/models/composer/composer.js

```diff
@@ -118,6 +120,31 @@ registerModel({
                 textInputSelectionDirection: clear(),
             });
         },
+        /**
+         * Try to fetch link preview when urls are inserted inside the composer.
+         */
+        getLinkPreview(pastedText) {
+            const urls = pastedText.match(urlRegexp) || false;
+            if (!urls) {
+                return;
+            }
+            // Add attachment for urls that are not already in attachment
+            const existingUrls = this.attachments.map(att => att.url);
+            const filteredUrls = urls.filter(url => !existingUrls.includes(url));
+            for (const url of filteredUrls) {
+                this.performRPClinkPreview(url);
+            }
+        },
+        async performRPClinkPreview(url) {
+            const attachment = await this.env.services.rpc({
+                route: '/mail/link_preview',
+                params: { url, channel_id: this.activeThread.id },
+            }, { shadow: true });
+            if (!attachment) {
+                return;
+            }
+            this.update({ attachments: insert(attachment) });
```

Comment:
Nit-picking but var name `attachment` is slightly misleading because it's not a attachment record but attachment data actually. Would add "Data" suffix to lift the ambiguity.